### PR TITLE
Removed the built-in overlay from ReorderableListView.

### LINKED ISSUE #76150: Inconsistency between ListView and ReorderableListView
Hi guys,
I'm trying to make a list of sections where each element has a reorderable list of children inside.

First attempt with `ListView `is ok, everything works as expected:

```
@override
  Widget build(BuildContext context) {
    return StreamBuilder(
      stream: _myBloc.sectionsStream,
      initialData: <Section>[],
      builder: (context, snapshot) {
        List<Section> sections = snapshot.data;
        return SingleChildScrollView(
          child: Column(
            children: List.generate(sections.length, (index) {
              Section section = sections.elementAt(index);
              return ListTile(
                title: Text(section.title),
                subtitle: ListView.builder(
                  shrinkWrap: true,
                  physics: ClampingScrollPhysics(),
                  itemCount: section.children.length,
                  itemBuilder: (context, index) {
                    Section child = section.children.elementAt(index);
                    return Card(
                      child: Text(child.title),
                    );
                  },
                ),
              );
            }),
          ),
        );
      },
    );
```
But then when I replace the `ListView `with a `ReorderableListView `with the same attributes, everything breaks:

```
@override
  Widget build(BuildContext context) {
    return StreamBuilder(
      stream: _organizeBloc.sectionsStream,
      initialData: <Section>[],
      builder: (context, snapshot) {
        List<Section> sections = snapshot.data;
        return SingleChildScrollView(
          child: Column(
            children: List.generate(sections.length, (index) {
              Section section = sections.elementAt(index);
              return ListTile(
                title: Text(section.title),
                subtitle: ReorderableListView.builder(
                  shrinkWrap: true,
                  physics: ClampingScrollPhysics(),
                  itemCount: section.children.length,
                  itemBuilder: (context, index) {
                    Section child = section.children.elementAt(index);
                    return Card(
                      child: Text(child.title),
                    );
                  },
                  onReorder: (oldIndex, newIndex) {},
                ),
              );
            }),
          ),
        );
      },
    );
```

Error returned is an incomprehensible: 

> The following assertion was thrown during performResize():
> 'package:flutter/src/widgets/overlay.dart': Failed assertion: line 720 pos 12: 'constraints.biggest.isFinite': is not true.

This seems to me an inconsistent behavior. Working with lists is quite a hell in Flutter compared to how it was with Android native... :(

### PR DESCRIPTION
**Pull request by darrenaustin**
The current implementation of `ReorderableListView` creates its own `Overlay` widget to handle the dragging of its items. However, this is causing layout problems for apps that are trying to use `shrinkWrap`, as the overlay is not constrained.
This PR removes this and has the ReorderableListView depend on the surrounding overlay provided by MaterialApp. If one isn't found it will throw an error.

Fixes: #76150

I added a new test to verify an error if no surrounding overlay is found. Also removed two golden tests that would need to be updated as the dragged item is no longer painting in the list itself (they were of no real testing value either).

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [x] All existing and new tests are passing.

### CODE REVIEW

File: packages/flutter/lib/src/material/reorderable_list.dart [hunk 1]
```diff
@@ -289,115 +289,7 @@ class ReorderableListView extends StatefulWidget {
   _ReorderableListViewState createState() => _ReorderableListViewState();
 }
 
-// This top-level state manages an Overlay that contains the list and
```
Comment (darrenaustin) [inline packages/flutter/lib/src/material/reorderable_list.dart:292]: This looks more drastic than it is. The previous implementation of _RerorderableListViewState wrapped  `_ReorderableListContent` with an overlay that it managed. This is no longer needed, so this just removes the `_ReorderableListContent` widget and replaces the _ReorderableListViewState with the contents of the old `_ReorderableListContentState` which is where the real list state implementation was.

File: packages/flutter/test/material/reorderable_list_test.dart [hunk 2]
```diff
@@ -212,47 +212,6 @@ void main() {
         expect(getListHeight(), kDraggingListHeight);
       });
 
-      testWidgets('Vertical drop area golden', (WidgetTester tester) async {
```
Comment (goderbauer) [inline packages/flutter/test/material/reorderable_list_test.dart:215]: What did these tests verify? Why do we no longer need to guarantee that this use case works?

**Review by goderbauer — COMMENTED** [review]
I thought there was a reason why we added the overlay (e.g. to make sure that you can't drag an item outside of the list). Why do we not need that anymore?

**Comment by darrenaustin** [comment]
> I thought there was a reason why we added the overlay (e.g. to make sure that you can't drag an item outside of the list). Why do we not need that anymore?

As far as I remember I had removed it, but we put it back near the end of development as it was causing some tests to fail or some other reason.

I don't think having a built-in overlay prevents the item from being dragged outside the list, as it doesn't appear to be clipping it (and I am not sure we would want it to from a UI perspective). The list already handles the case of dragging outside the list and just puts the gap at the edge of the list and the item will animate back to that gap if dropped outside the list.

In addition, it is inconsistent with the other two reorderable list widgets as they both rely on the surrounding overlay widget.

Therefore I don't see much value in keeping the build-in overlay here as it provides little value and is breaking the shrink-wrap case as it doesn't shrink wrap around the list. 

Or am I missing something here?

File: packages/flutter/test/material/reorderable_list_test.dart (hunk 2 above)
Comment (darrenaustin) [reply packages/flutter/test/material/reorderable_list_test.dart:215]: It is not entirely clear from the code, but from the title 'Vertical drop area golden', I presume it was trying to test that there was a drop area when an item was dragged. However, looking at this code it creates a three item list, starts dragging the middle item (blue) and then takes a screen shot. However, as the item has not yet moved, it is sitting over the drop area, so the golden image is just a list with three items in it. Not sure what it was trying to verify here.  Removing the overlay would mean the item is no longer in the list view's repaint boundary, so the golden would have to be updated, but again, I am not sure what value it provides.

**Comment by goderbauer** [comment]
> I don't think having a built-in overlay prevents the item from being dragged outside the list, as it doesn't appear to be clipping it (and I am not sure we would want it to from a UI perspective).

I thought there was a comment in the code explaining why there is an internal Overlay, but looks like it got lost during the refactor. Prior to the refactor, this is what it said:

https://github.com/flutter/flutter/blob/ff05ca2503437ca35ea07551aeda5a6b1d5d35bb/packages/flutter/lib/src/material/reorderable_list.dart#L175

And looking at the Overlay I would expect it to be clipping (it's the default value for https://master-api.flutter.dev/flutter/widgets/Overlay/clipBehavior.html and we don't seem to change it...). Is that not working?

Whether not drawing outside the list itself is a bug or a feature, I actually don't know... The more I think about it the more I am leaning towards it should draw outside of the list. So, if that was the only reason, maybe we should remove it...

File: packages/flutter/test/material/reorderable_list_test.dart (hunk 2 above)
Comment (goderbauer) [reply packages/flutter/test/material/reorderable_list_test.dart:215]: > However, as the item has not yet moved, it is sitting over the drop area, so the golden image is just a list with three items in it.

Is that maybe what the test was testing? If you pick up an item it starts out to render at the correct location (i.e. at its original location in the list).

**Comment by darrenaustin** [comment]
> And looking at the Overlay I would expect it to be clipping (it's the default value for https://master-api.flutter.dev/flutter/widgets/Overlay/clipBehavior.html and we don't seem to change it...). Is that not working?
> 

I just tried it and it seemed to paint the dragged item over the app bar and such just fine. I will take another look to see if I was just holding wrong or something.

> Whether not drawing outside the list itself is a bug or a feature, I actually don't know... The more I think about it the more I am leaning towards it should draw outside of the list. So, if that was the only reason, maybe we should remove it...

That's pretty much the conclusion that led me to this PR. If you do come up with a case for the built-in overlay, just let me know.

File: packages/flutter/test/material/reorderable_list_test.dart (hunk 2 above)
Comment (darrenaustin) [reply packages/flutter/test/material/reorderable_list_test.dart:215]: Yeah, as I was explaining this above, that thought occurred to me as well. I will update the test to show the missing item during the drag and update the golden for it. Thx for the input.

**Comment by darrenaustin** [comment]
> I just tried it and it seemed to paint the dragged item over the app bar and such just fine. I will take another look to see if I was just holding wrong or something.

Ok, I checked again, and I was holding it wrong. It does clip, but as mentioned before I don't think that is the behavior we want here. It is inconsistent with the other reorderable list widgets and if someone really wants the dragged item to clip, they can wrap it with their own overlay.

File: packages/flutter/test/material/reorderable_list_test.dart (hunk 2 above)
Comment (darrenaustin) [reply packages/flutter/test/material/reorderable_list_test.dart:215]: I just pushed a commit that re-added these tests, but updated them for capturing a drag in progress.

**Comment by skia-gold** [comment]
Gold has detected about 6 untriaged digest(s) on patchset 2.
View them at https://flutter-gold.skia.org/cl/github/79024

File: packages/flutter/test/material/reorderable_list_test.dart [hunk 3]
```diff
@@ -212,7 +212,8 @@ void main() {
...
+      testWidgets('Vertical drag in progress golden image', (WidgetTester tester) async {
+        debugDisableShadows = false;
```
Comment (goderbauer) [inline packages/flutter/test/material/reorderable_list_test.dart:216]: I believe this has to be reset to its original value at the end of the test?

File: packages/flutter/test/material/reorderable_list_test.dart [hunk 4]
```diff
@@ -1342,6 +1376,38 @@ void main() {
     expect(exception, isFlutterError);
     expect(exception.toString(), contains('Every item of ReorderableListView must have a key.'));
   });
+
```
Comment (goderbauer) [inline packages/flutter/test/material/reorderable_list_test.dart:1379]: Should we also include a test where the Overlay does not tightly wrap the list (i.e. the list has to share the screen space with another sibling) to ensure that our math is correct to the origin of the overlay and not the origin of the list?

**Review by goderbauer — APPROVED** [review]
LGTM

**Comment by goderbauer** [comment]
> as mentioned before I don't think that is the behavior we want here. It is inconsistent with the other reorderable list widgets and if someone really wants the dragged item to clip, they can wrap it with their own overlay.

Sounds good to me.

**Pull request merged**